### ISSUE TITLE
SNI support in Mojo::Server::Daemon

### ISSUE THREAD

**Issue by kraih** [kraih/mojo#882]
We already have SNI support in `Mojo::UserAgent`, but for HTTP/2 it might also be nice to have it in `Mojo::Server::Daemon`. Some creative use of query parameters in the listen locations should make for a reasonable API, like `listen => ['https://*:443?mojolicio.us_cert=foo.crt&mojolicio.us_key=foo.key&mojolicious.org_cert=bar.cert&mojolicious.org_key=bar.key']`.

**Comment by kraih** [comment]
This feature might depend on #876.

**Comment by kraih** [comment]
We basically need to pass hashes to `SSL_cert_file` and `SSL_key_file` in `Mojo::IOLoop::Server`.
```
SSL_cert_file => {
  'mojolicio.us'    => 'foo.crt',
  'mojolicious.org' => 'bar.crt'
},
SSL_key_file => {
  'mojolicio.us'    => 'foo.key',
  'mojolicious.org' => 'bar.key'
}
```

**Comment by kraih** [comment]
Testing all of this might require a little more creativity though.

**Comment by scottw** [comment]
Fwiw, I've written a PoC server that create two certs each with a different hostname. I start up a server using Mojo::IOLoop->server directly using the hashref syntax:

        tls_cert   => { 'host1.example.com' => 'host1.crt', 'host2.example.com' => 'host2.crt', '' => 'host1.crt' },
        tls_key    => { 'host1.example.com' => 'host1.key', 'host2.example.com' => 'host2.key', '' => 'host1.key' },

and it correctly serves the right certificate based on the connecting hostname. I'm using the openssl s_client to test this. Testing using Perl tools is proving a formidable challenge, in no small part due to the fact that IO::Socket::SSL's "can_server_sni()" only does an OpenSSL version comparison, instead of a true feature test for SNI negotiation.

At any rate, if someone is eager for this feature to be done in a hurry, I'd be delighted to send along my PoC work. In the meantime, I suspect I'll be spending some quality time with Perl's lower level SSL libraries. No ETA.

**Comment by s1037989** [comment]
I think this is very important.  It's a very real-world scenario.  Fortunately nginx as a reverse proxy in front of Mojolicious works perfectly well, but it would be nice to not need to stand up an nginx instance just for that.  I'm perfectly happy running traffic straight to Mojolicious.

Currently, I use Let's Encrypt to generate my free certs and that works well.  I run Mojolicious behind Toadfarm and mount many different apps.  I have to list all CNs in the cert.  That's not only annoying, but all CNs for the cert are visible which I don't think is cool.

**Comment by kraih** [comment]
Testing this might actually be a lot easier than i though, with a fake proxy server like the one we use in `t/mojo/websocket_proxy_tls.t`.

**Comment by kraih** [comment]
And implemented. https://github.com/kraih/mojo/commit/a2469e62196573bf8d60822e07257daac79b4dd8

**Comment by s1037989** [comment]
:+1:  :dancer:
